### PR TITLE
support async functions as auth functions

### PR DESCRIPTION
**Pull request by PavelPolyakov**
support async functions as auth functions;
remove deprecation warning;

resolves https://github.com/fastify/fastify-auth/issues/35

@Eomm, please, have a look

Regards,

### CODE REVIEW

File: test/async-test.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+const test = t.test
+
+const fastify = require('fastify')()
+
+fastify
+  .decorate('verifyJWTandLevel', function (request, reply) {
+    return new Promise(function (resolve, reject) {
+      if (request.body && request.body.failure) {
+        reply.code(401).send({ 'error': 'Unauthorized' })
+        reject(new Error())
```
Comment (Eomm) [inline test/async-test.js]: I think here we shouldn't use the reply.send to send the error, but only reject with an error `reject(new Error('Password not valid')` and check that the output will be a status code 401

File: test/async-test.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
+'use strict'
+
+const t = require('tap')
+const test = t.test
+
+const fastify = require('fastify')()
```
Comment (Eomm) [inline test/async-test.js]: We could copy the same logic of the other test using the `const build = require('./example')` but we could create a new file `example-async.js` that is equal to `example` but use the promise instead with the same tests. It should be simple and so we can cover all the cases.

And of course, we should also test the mixed case (one promise and one sync function). The edge case 💪

File: test/test.js [hunk 3]
```diff
@@ -89,26 +89,6 @@ test('Auth succesful', t => {
   })
 })
 
-test('Auth not succesful', t => {
```
Comment (Eomm) [inline test/test.js]: Why this test has been removed?

**Comment by PavelPolyakov** [comment]
Big thanks for the good review, will adjust PR tomorrow.--Отправлено из мобильной Яндекс.Почты18.03.2019, 22:10, "Manuel Spigolon" <notifications@github.com>:@Eomm requested changes on this pull request.



In fastify-auth.js:
> @@ -59,7 +59,11 @@ function auth (functions) {
         return
       }
 
-      func(that.request, that.reply, that.onAuth)
+      var maybePromise = func(that.request, that.reply, that.onAuth)
+
+      if (Promise.resolve(maybePromise) === maybePromise) {

To understand if the result is a promise you can check like this:
⬇️ Suggested change
-      if (Promise.resolve(maybePromise) === maybePromise) {
+     if (maybePromise && typeof maybePromise.then === 'function')


This will not create more promise in memory so the garbage collection will be more relaxed.



In fastify-auth.js:
> @@ -59,7 +59,11 @@ function auth (functions) {
         return
       }
 
-      func(that.request, that.reply, that.onAuth)
+      var maybePromise = func(that.request, that.reply, that.onAuth)
+
+      if (Promise.resolve(maybePromise) === maybePromise) {
+        maybePromise.then(that.onAuth).catch(that.onAuth)

this could be error prone: if the promise return { someThing: 42 } the that.onAuth will use that object like an error.
So you can write
⬇️ Suggested change
-        maybePromise.then(that.onAuth).catch(that.onAuth)
+        maybePromise.then(results => that.onAuth(null, results), that.onAuth)





In package.json:
> @@ -10,7 +10,7 @@
   "types": "fastify-auth.d.ts",
   "scripts": {
     "clean": "rimraf authdb",
-    "test": "standard && tap test.js",
+    "test": "standard && tap test",

Could you write && tap -J test/*.test.js so this would be more readable?



In test/async-test.js:
> @@ -0,0 +1,62 @@
+'use strict'
+
+const t = require('tap')
+const test = t.test
+
+const fastify = require('fastify')()
+
+fastify
+  .decorate('verifyJWTandLevel', function (request, reply) {
+    return new Promise(function (resolve, reject) {
+      if (request.body && request.body.failure) {
+        reply.code(401).send({ 'error': 'Unauthorized' })
+        reject(new Error())

I think here we shouldn't use the reply.send to send the error, but only reject with an error reject(new Error('Password not valid') and check that the output will be a status code 401



In test/async-test.js:
> @@ -0,0 +1,62 @@
+'use strict'
+
+const t = require('tap')
+const test = t.test
+
+const fastify = require('fastify')()

We could copy the same logic of the other test using the const build = require('./example') but we could create a new file example-async.js that is equal to example but use the promise instead with the same tests. It should be simple and so we can cover all the cases.
And of course, we should also test the mixed case (one promise and one sync function). The edge case <g-emoji class="g-emoji" alias="muscle" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4aa.png">💪



In test/test.js:
> @@ -89,26 +89,6 @@ test('Auth succesful', t => {
   })
 })
 
-test('Auth not succesful', t => {

Why this test has been removed?

File: test/async-test.js (hunk 1 above)
Comment (PavelPolyakov) [reply test/async-test.js]: done

File: test/test.js (hunk 3 above)
Comment (PavelPolyakov) [reply test/test.js]: was not meant to, thanks for pointing out

File: test/async-test.js (hunk 2 above)
Comment (PavelPolyakov) [reply test/async-test.js]: Agreed, I've based new tests on top of new `example-async.js` and updated both examples with explicit `reply` call from the auth function.

File: example-async.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      if (request.body && request.body.failureWithReply) {
+        reply.code(401).send({ 'error': 'Unauthorized' })
+        return reject(new Error())
+      }
+
+      if (!request.req.headers['auth']) {
+        return reject(new Error('Missing token header'))
+      }
+
+      jwt.verify(request.req.headers['auth'], onVerify)
```
Comment (mcollina) [inline example-async.js]: I would use `util.promisify` to get a promise-based version of this.

File: example-async.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    .after(routes)
+
+  fastify.decorate('verifyJWTandLevel', verifyJWTandLevel)
+  fastify.decorate('verifyUserAndPassword', verifyUserAndPassword)
+
+  function verifyJWTandLevel (request, reply) {
+    const jwt = this.jwt
+    const level = this.level
+
+    return new Promise(function (resolve, reject) {
```
Comment (mcollina) [inline example-async.js]: I would not use `new Promise()` here.

Comment (PavelPolyakov) [reply example-async.js]: As for me the idea of the test is to test what if `auth` function returns `Promise`.
How exactly it returns it is out of scope.

I don't mind changing this based on your suggestion though.

File: example-async.js (hunk 4 above)
Comment (PavelPolyakov) [reply example-async.js]: @mcollina can you, please, check if this is what you meant?
https://github.com/PP-etc/fastify-auth/blob/feature/refactor-to-use-util.promisify/example-async.js#L31

To me mix of callback and Promise looks worse, than just returning `Promise` right away and using `resolve/reject` after. 

But I believe I've solved it not in the best possible way, please, point me what can be improved.

Comment (mcollina) [reply example-async.js]: To me the full `verifyJWTandLevel` function needs to become an async function. 

```js
const decoded = await util.promisify(jwt.verify)(request.req.headers['auth'])
if (!decoded.user || !decoded.password) {
  throw new Error('Token not valid')
}
try {
  const user = await level.get(decoded.user)
 // ...
} catch (err) {
   // ..
}
```

Comment (PavelPolyakov) [reply example-async.js]: I use `Promise` because there are tests for `node 6`: https://travis-ci.org/fastify/fastify-auth/jobs/508269248 .

If it's okay to drop it of course proper `async/await` will look better.

Comment (mcollina) [reply example-async.js]: You can use promises all the way.

Comment (PavelPolyakov) [reply example-async.js]: https://github.com/PP-etc/fastify-auth/blob/feature/refactor-to-use-util.promisify/example-async.js#L18

please, check if this is what you mean

File: example-async.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  fastify
+    .register(require('fastify-jwt'), { secret: 'supersecret' })
+    .register(require('fastify-leveldb'), { name: 'authdb-async' })
+    .register(require('./fastify-auth'))
+    .after(routes)
+
+  fastify.decorate('verifyJWTandLevel', verifyJWTandLevel)
+  fastify.decorate('verifyUserAndPassword', verifyUserAndPassword)
+
+  function verifyJWTandLevel (request, reply, done) {
```
Comment (mcollina) [inline example-async.js]: Do not mix promises and callbacks.

File: example-async.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    const jwt = this.jwt
+    const level = this.level
+
+    if (request.body && request.body.failureWithReply) {
+      reply.code(401).send({ error: 'Unauthorized' })
+      return done(new Error())
```
Comment (mcollina) [inline example-async.js]: I think it’s better if you do a return Promise.reject(err). If you use the Unhauthorized error class from http-errors, there is no need to call reply.code(401).

File: example-async.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    return util
+      .promisify(jwt.verify)(request.req.headers['auth'])
+      .then(function (decoded) {
+        if (!decoded.user || !decoded.password) {
+          throw new Error('Token not valid')
+        }
+        util
+          .promisify(level.get).bind(level)(decoded.user)
```
Comment (mcollina) [inline example-async.js]: you should return this promise to maintain the chain.

level should already support promises, can you check?

File: example-async.js [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+            if (!password || password !== decoded.password) {
+              throw new Error('Token not valid')
+            }
+          })
+          .catch(function (err) {
+            if (err.notFound) {
+              throw new Error('Token not valid')
+            }
+            throw err
+          })
```
Comment (mcollina) [inline example-async.js]: I think you can condense this into the external .catch block.

File: example-async.js (hunk 8 above)
Comment (PavelPolyakov) [reply example-async.js]: yes, that works, thanks

File: example-async.js (hunk 6 above)
Comment (PavelPolyakov) [reply example-async.js]: agreed, `Promise.reject` was what I've missed

File: example-async.js (hunk 7 above)
Comment (PavelPolyakov) [reply example-async.js]: regarding the `http-errors` - do you mean I should add https://www.npmjs.com/package/http-errors as dependency?

File: example-async.js (hunk 9 above)
Comment (PavelPolyakov) [reply example-async.js]: Can you, please, elaborate? For now I don't see how can this be simplified

File: example-async.js (hunk 4 above)
Comment (PavelPolyakov) [reply example-async.js]: I've replaced `util.promisify` with `new Promise`

File: example-async.js [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        .then(function (password) {
+          if (!password || password !== decoded.password) {
+            throw new Error('Token not valid')
+          }
+        })
+        .catch(function (err) {
+          if (err.notFound) {
+            throw new Error('Token not valid')
+          }
+          throw err
```
Comment (mcollina) [inline example-async.js]: This err is going to be caught by the external catch,  and it will never be surfaced. That’s ok, but that means we can remove this catch completely.

Comment (PavelPolyakov) [reply example-async.js]: got you, thanks, done

File: fastify-auth.js [hunk 11]
```diff
@@ -59,7 +59,11 @@ function auth (functions) {
         return
       }
 
-      func(that.request, that.reply, that.onAuth)
+      var maybePromise = func(that.request, that.reply, that.onAuth)
+
+      if (maybePromise && typeof maybePromise.then === 'function') {
+        maybePromise.then(results => that.onAuth(null, results)).catch(that.onAuth)
```
Comment (delvedor) [inline fastify-auth.js]: Can you pass both callbacks to `.then`?
```suggestion
        maybePromise.then(results => that.onAuth(null, results), that.onAuth)
```

File: example-async.js [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+      return Promise.reject(new Error())
+    }
+
+    if (!request.req.headers['auth']) {
+      return Promise.reject(new Error('Missing token header'))
+    }
+
+    return new Promise(function (resolve, reject) {
+      jwt.verify(request.req.headers['auth'], function (err, decoded) {
+        if (err) { reject(err) };
```
Comment (Eomm) [inline example-async.js]: We should add a return after the reject otherwise we will cause a memory leak (and we don't want to teach something wrong to users 😁)

File: example-async.js [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+      })
+    }).then(function (decoded) {
+      return level.get(decoded.user)
+        .then(function (password) {
+          if (!password || password !== decoded.password) {
+            throw new Error('Token not valid')
+          }
+        })
+    }).catch(function () {
+      throw new Error('Token not valid')
```
Comment (Eomm) [inline example-async.js]: (missing the err parameter)
```suggestion
      throw new Error(err.message)
```

Comment (mcollina) [reply example-async.js]: I think using “Token not valid” is a security feature and it should be retained

Comment (Eomm) [reply example-async.js]: Of course, but the error is thrown in a "then" so we don't lost this behavior, we only add a check that there are not other masquerade errors. 
This was my intent here

Comment (mcollina) [reply example-async.js]: We can log the error and rethrow. The errors are purposefully masked so that they don't baloon in the Fastify error handler and then displayed to the user.

File: example-async.js (hunk 12 above)
Comment (PavelPolyakov) [reply example-async.js]: done, thanks

File: example-async.js [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+if (require.main === module) {
+  const fastify = build({
+    logger: {
+      level: 'info'
+    }
+  })
+  fastify.listen(3000, err => {
+    if (err) throw err
+    console.log(
+      `Server listenting at http://localhost:${fastify.server.address().port}`
```
[the remark this answers is not in the document]
Comment (PavelPolyakov) [reply example-async.js]: done

File: test/example.test.js [hunk 15]
```diff
@@ -3,7 +3,7 @@
 const t = require('tap')
 const test = t.test
 const rimraf = require('rimraf')
-const build = require('./example')
+const build = require('../example')
```
[the remark this answers is not in the document]
Comment (PavelPolyakov) [reply test/example.test.js]: done, thanks

File: example-async.js (hunk 7 above)
Comment (PavelPolyakov) [reply example-async.js]: given that it seems not to be a requirement and such small use case it not yet enough for me to add another dev dependency - I'll just resolve this discussion

File: example-async.js (hunk 13 above)
Comment (PavelPolyakov) [reply example-async.js]: Current state is kind of a successor of this approach: https://github.com/fastify/fastify-auth/blob/master/example.js#L29 , so there is not real error but generic one.

Added the log before the error is thrown.

File: fastify-auth.js (hunk 11 above)
Comment (PavelPolyakov) [reply fastify-auth.js]: thanks, updated

**Comment by Eomm** [comment]
@mcollina sure 👍
